components/server/src/api/auth-provider-service-api.ts

```diff
@@ -0,0 +1,189 @@
+        }
+    }
+    async getAuthProvider(request: GetAuthProviderRequest, context: HandlerContext): Promise<GetAuthProviderResponse> {
+        if (!request.authProviderId) {
+            throw new ConnectError("authProviderId is required", Code.InvalidArgument);
+        }
+
+        const authProvider = await this.authProviderService.getAuthProvider(context.user.id, request.authProviderId);
+        if (!authProvider) {
+            throw new ConnectError("Provider not found.", Code.NotFound);
+        }
+
+        return new GetAuthProviderResponse({
+            authProvider: this.apiConverter.toAuthProvider(authProvider),
+        });
+    }
+
+    async listAuthProviders(
+        request: ListAuthProvidersRequest,
+        context: HandlerContext,
+    ): Promise<ListAuthProvidersResponse> {
+        const target = request.id;
+        const ownerId = target.case === "userId" ? target.value : undefined;
+        const organizationId = target.case === "organizationId" ? target.value : undefined;
+
+        if (!organizationId && !ownerId) {
+            throw new ConnectError("organizationId or ownerId is required", Code.InvalidArgument);
+        }
+
+        const authProviders = organizationId
+            ? await this.authProviderService.getAuthProvidersOfOrg(context.user.id, organizationId)
+            : await this.authProviderService.getAuthProvidersOfUser(context.user.id);
+
+        const redacted = authProviders.map(AuthProviderEntry.redact.bind(AuthProviderEntry));
+
+        const result = new ListAuthProvidersResponse({
+            authProviders: redacted.map((ap) => this.apiConverter.toAuthProvider(ap)),
+        });
+        return result;
+    }
+
+    /**
+     * Listing descriptions of auth providers doesn't require authentication.
+     */
+    @Unauthenticated()
+    async listAuthProviderDescriptions(
+        _request: ListAuthProviderDescriptionsRequest,
+        context: HandlerContext,
+    ): Promise<ListAuthProviderDescriptionsResponse> {
+        const user = context.user;
+        const aps = user
+            ? await this.authProviderService.getAuthProviderDescriptions(user)
+            : await this.authProviderService.getAuthProviderDescriptionsUnauthenticated();
+
+        return new ListAuthProviderDescriptionsResponse({
+            descriptions: aps.map((ap: AuthProviderInfo) => this.apiConverter.toAuthProviderDescription(ap)),
+        });
+    }
+
+    async updateAuthProvider(
+        request: UpdateAuthProviderRequest,
+        context: HandlerContext,
+    ): Promise<UpdateAuthProviderResponse> {
+        if (!request.authProviderId) {
+            throw new ConnectError("authProviderId is required", Code.InvalidArgument);
+        }
+        const clientId = request.clientId;
+        const clientSecret = request.clientSecret;
+        if (!clientId || typeof clientSecret === "undefined") {
```

Comment:
we don't require both? it is rather at least one of them should be present

Reply:
I used following pattern here: https://github.com/gitpod-io/gitpod/blob/c33b555448b0b566600282615402fb94cd72dd92/components/server/src/api/organization-service-api.ts#L244-L265

not sure whether we want to align

Reply:
This would mean, we need to change the Dashboard part and retest the behavior. I'd rather avoid.

Reply:
- We don't build API for one client (dashboard).
- If dashboard already provides both values it should not an issue anyway?

Reply:
Ah, now I got you. 

> we don't require both? it is rather at least one of them should be present

wasn't clear from that. 

So, even if Dashboard is always sending the tuple, we may allow for adjusting a single attribute here as well, right? We can change that. Just the focus wasn't on changing existing behavior for the migration phase.

Reply:
resolved in fc3be827450d6a0a4f3c00649b903f4e492cd284